### PR TITLE
Add hydraulic design-verification extension

### PR DESCRIPTION
**Pull request by thom1505**
## Summary

This pull request introduces a hydraulic design-verification extension for WNTR.

The extension provides a structured workflow for applying proposed pipe-diameter designs and hydraulic operating scenarios to independent copies of a WNTR water network model and verifying their hydraulic feasibility.

## Main capabilities

- Applies proposed pipe-diameter changes without modifying the original network model.
- Configures hydraulic operating scenarios for verification.
- Runs independent WNTR hydraulic simulations.
- Evaluates minimum junction pressure.
- Evaluates maximum absolute pipe velocity.
- Calculates pressure and velocity compliance percentages.
- Identifies critical nodes, pipes and simulation times.
- Produces an overall hydraulic feasibility result.
- Records audit information describing the design and scenario changes applied.
- Supports verification of pump-related hydraulic conditions where applicable.

## Testing

Automated tests were developed to assess:

- pressure and velocity constraint calculations;
- pipe-design application;
- hydraulic-scenario configuration;
- preservation of the original network model;
- hydraulic verification outputs;
- feasibility decisions;
- audit and diagnostic information; and
- pump-related verification behaviour.

The relevant tests passed locally before submission.

## Purpose

The contribution is intended to support transparent and reproducible verification of proposed water-distribution-network designs while retaining the WNTR hydraulic simulator as the independent verification layer.

## Notes for reviewers

This pull request is submitted for technical review and feedback. I would welcome recommendations regarding API design, naming, documentation, testing coverage and alignment with WNTR contribution standards.

### CODE REVIEW

**Comment by thom1505** [comment]
Hello, the GitHub Actions workflows for this pull request are currently awaiting maintainer approval. The extension test suite has been run locally on Python 3.12.13, with all 40 tests passing, including the additional pump-curve, incomplete-data, and reverse-flow regression tests. Could a maintainer please approve and run the workflows when convenient? Thank you.

**Comment by thom1505** [comment]
I have pushed additional commits addressing pressure-exponent configuration, pump-curve edge cases, invalid pump-status observations, and pump diagnostics in batch summaries. The complete local extension test suite currently passes 49 tests. The GitHub workflows are showing “completed with no jobs” and may require maintainer approval before they can run.

**Comment by thom1505** [comment]
Additional validation has now been completed locally. The hydraulic design-verification extension passes 53 extension tests, including a direct EPANET 2.2 integration test. The broader WNTR quick-test suite also completed with 466 tests passing, 6 skipped, and no failures. The documentation HTML build completed successfully, the files changed by this pull request contain no suspicious Unicode characters, and the

**Comment by thom1505** [comment]
I have continued strengthening the extension with additional fail-safe validation, regression testing and documentation updates. The extension test suite passes locally. I noticed that the GitHub Actions workflows are awaiting maintainer approval because the pull request originates from a fork. When convenient, I would appreciate approval of the workflow runs so that the upstream checks can execute.

File: documentation/extensions/hydraulic_design_verification.rst [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+   >>> scenario.name
+   'Baseline DD'
+   >>> scenario.demand_model
+   'DD'
+
+The following example assumes that ``wn`` is an existing
+``WaterNetworkModel``:
+
+.. code-block:: python
+
```
Comment (kaklise) [inline documentation/extensions/hydraulic_design_verification.rst:71]: This code block should be a doctest that includes a water network model.  The print statements at the end should include an explanation of results.

File: documentation/extensions/hydraulic_design_verification.rst [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+``IncompleteHydraulicResultsError`` remains a subclass of
+``RuntimeError``. Existing applications that catch these standard
+exception types therefore remain compatible. Incorrect Python object
+types continue to raise ``TypeError``.
+
+Applications can catch either the common base exception or a specific
+failure type:
+
+.. code-block:: python
+
```
Comment (kaklise) [inline documentation/extensions/hydraulic_design_verification.rst]: Code block should be a doctest, or remove the focus on specific error types and use simple assert statements to check for expectations.  This could help focus the extension on the analysis.

File: documentation/extensions/hydraulic_design_verification.rst [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+Hydraulic Design Verification
+=============================
+
+**Summary:** The ``design_verification`` extension applies proposed pipe
+designs and hydraulic operating scenarios to independent copies of a
+WNTR water-distribution network model. It runs a hydraulic simulation
+and checks whether pressure, velocity and pump operating requirements
+are satisfied.
+
```
Comment (kaklise) [inline documentation/extensions/hydraulic_design_verification.rst:9]: The documentation needs to clarify that the design verification considers pipe diameters and simulation options (namely, time and hydraulic options).  Additional pipe parameters and operating conditions (including controls) are not included in the verification.

File: wntr/extensions/design_verification/models.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        Minimum-pressure constraint result.
+    velocity_result
+    Maximum-velocity constraint result.
+    pump_result
+    Head-pump operating-curve audit result, when available.
+    feasible
+    Whether pressure, velocity and pump operating-curve
+    requirements were met.
+    simulation_error_code
+    Simulator error code, when one is supplied by WNTR.
```
Comment (kaklise) [inline wntr/extensions/design_verification/models.py]: check doc string formatting throughout

File: wntr/extensions/design_verification/runner.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+
+    normalized = value.strip().upper()
+
+    aliases = {
+        "WNTR": "WNTR",
+        "WNTRSIMULATOR": "WNTR",
+        "WNTR SIMULATOR": "WNTR",
+        "EPANET": "EPANET",
+        "EPANETSIMULATOR": "EPANET",
+        "EPANET SIMULATOR": "EPANET",
```
Comment (kaklise) [inline wntr/extensions/design_verification/runner.py]: I suggest only using the simulator names from WNTR, EpanetSimulator or WNTRSimulator

File: wntr/extensions/design_verification/models.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    diameters_m
+        Mapping of pipe names to proposed diameters in metres.
+    """
+
+    name: str
+    diameters_m: Mapping[str, float]
+
+
+@dataclass(frozen=True)
+class HydraulicScenario:
```
Comment (kaklise) [inline wntr/extensions/design_verification/models.py:58]: I suggest reusing the wntr.network.Options class instead of creating a new one.  This will also make it clear that the Scenario is changing simulation options.  You can still create a wrapper around the options class to apply the changes. This will use built in error checking for things like valid string/float/int that will simplify the code in this PR.  For example:

Define the simulation settings for the scenario
```
options = wntr.network.Options()
options.time.duration = 24*3600
options.time.hydraulic_timestep = 3600
options.time.report_timestep = 3600
options.hydraulic.demand_model = 'PDD'
options.hydraulic.demand_multiplier = 1.25
options.hydraulic.required_pressure = 15
options.hydraulic.pressure_exponent = 0.5
```
And apply the changes (simplifies `apply_hydraulic_scenario`)
```
scneario_wn = copy.deepcopy(wn)
scneario_wn.options.time = options.time
scneario_wn.options.hydraulic = options.hydraulic
```

File: wntr/extensions/design_verification/__init__.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
+"""Hydraulic design-verification extension for WNTR.
+
+The extension applies proposed pipe designs and hydraulic operating
```
Comment (kaklise) [inline wntr/extensions/design_verification/__init__.py]: Update text to clarify that the validation considers pipe diameter and hydraulic simulation options

File: wntr/extensions/design_verification/batch.py [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+        "network_hash": network_hash,
+        "status": None,
+        "started_at_utc": started_at_utc,
+        "elapsed_s": elapsed_s,
+        "design_name": (
+            None
+            if design is None
+            else _safe_name(design.name)
+        ),
+        "scenario_name": _safe_name(scenario.name),
```
Comment (kaklise) [inline wntr/extensions/design_verification/batch.py]: Remove the _safe_name function and use .strip directly.

File: wntr/extensions/design_verification/batch.py [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+        Detailed records keyed by experiment ID. Successful records
+        include verification and audit objects. Raw hydraulic results
+        are included only when ``retain_hydraulic_results`` is true.
+    """
+    if not isinstance(wn, WaterNetworkModel):
+        raise TypeError("wn must be a WNTR WaterNetworkModel.")
+
+    design_values = _normalise_designs(designs)
+    scenario_values = _normalise_scenarios(scenarios)
+    simulator_values = _normalise_simulators(simulators)
```
Comment (kaklise) [inline wntr/extensions/design_verification/batch.py]: Consider how the _normalise functions could be simplified using an assert statement, and then convert str to [str] if needed.  
```
assert isinstance(data, (str, list)) and (
      isinstance(data, str) or all(isinstance(item, str) for item in data)
  ), 'Input must be a string or a list of strings'
```

File: wntr/extensions/design_verification/pumps.py [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+    if not isinstance(flowrate, pd.DataFrame):
+        raise TypeError("flowrate must be a pandas DataFrame")
+
+    try:
+        tolerance = float(relative_tolerance)
+    except (TypeError, ValueError) as error:
+        raise InvalidConstraintError(
+            "relative_tolerance must be a finite "
+            "non-negative value"
+        ) from error
```
Comment (kaklise) [inline wntr/extensions/design_verification/pumps.py]: WNTR includes functions that check value type, see https://github.com/USEPA/WNTR/blob/main/wntr/utils/check_values.py.  This would simplify code like this throughout the extension.

File: wntr/extensions/design_verification/constraints.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+import math
+
+import numpy as np
+import pandas as pd
+
+from .exceptions import InvalidConstraintError
+from .models import ConstraintResult
+
+
+def _validate_result_table(
```
Comment (kaklise) [inline wntr/extensions/design_verification/constraints.py]: Can the extension rely on the construction of SimulationResults to validate that the results table exists?  For example, if the simulation runs to completion, the simulation results will be populated.

File: wntr/extensions/design_verification/runner.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+
+    if normalized not in aliases:
+        raise UnsupportedSimulatorError(
+            "simulator must be WNTR or EPANET."
+        )
+
+    return aliases[normalized]
+
+
+def _select_result_columns(
```
Comment (kaklise) [inline wntr/extensions/design_verification/runner.py]: It seems like this function could be simpler or eliminated and rely on pandas to extract the DataFrame column.  WNTR includes checks for duplicate node and link names in the registry.  If there are cases where the table  raises an error, then this is needed.

**Review by kaklise — COMMENTED** [review]
@thom1505 Thank you for submitting a WNTR extension! I have provided some feedback below.  Overall, I think the code should be simplified to help users understand the analysis. For example, the `HydraulicScenario` class is a subset of `wntr.network.Options`, and `wntr.utils.check_values` and other assert statements can be used throughout to simplify error checking.  I think these changes will greatly reduce the amount of code and help focus the effort on the analysis.

**Comment by thom1505** [comment]
Noted with thanks

File: documentation/extensions/hydraulic_design_verification.rst (hunk 1 above)
Comment (thom1505) [reply documentation/extensions/hydraulic_design_verification.rst:71]: Addressed. The example has been converted to an executable doctest that constructs a small water network model, applies the design-verification workflow, and reports explanatory pressure, velocity, and overall feasibility results. The documentation doctests pass successfully.

File: documentation/extensions/hydraulic_design_verification.rst (hunk 3 above)
Comment (thom1505) [reply documentation/extensions/hydraulic_design_verification.rst:9]: Addressed. The documentation now explicitly states that verification is limited to pipe diameters and the WNTR time and hydraulic option groups, and that additional pipe parameters and operating conditions, including controls, are not included.

File: wntr/extensions/design_verification/models.py (hunk 4 above)
Comment (thom1505) [reply wntr/extensions/design_verification/models.py]: Addressed. The NumPy-style docstrings were standardized throughout the design verification extension by adding parameter and return types consistently across the public data models and analysis functions. The full extension test suite passes with 68 tests.

File: wntr/extensions/design_verification/runner.py (hunk 5 above)
Comment (thom1505) [reply wntr/extensions/design_verification/runner.py]: Addressed. Simulator selection now uses only the WNTR simulator names WNTRSimulator and EpanetSimulator. The previous aliases were removed during the simplification of the verification runner.

File: wntr/extensions/design_verification/batch.py (hunk 8 above)
Comment (thom1505) [reply wntr/extensions/design_verification/batch.py]: Addressed. The _safe_name helper was removed, and .strip() is now used directly where design, scenario, and simulator names are handled.

File: wntr/extensions/design_verification/batch.py (hunk 9 above)
Comment (thom1505) [reply wntr/extensions/design_verification/batch.py]: Addressed. The _normalise_* helper functions were removed. The batch runner now converts the supplied inputs directly to lists as needed and uses simple assertions to validate design, scenario, and simulator values.

File: wntr/extensions/design_verification/pumps.py (hunk 10 above)
Comment (thom1505) [reply wntr/extensions/design_verification/pumps.py]: Addressed. The extension now reuses WNTR’s built-in value checking helpers in the pump audit, hydraulic-constraint validation, and pipe-diameter validation. Manual float and positivity checks were simplified while retaining extension-specific finite and range checks where needed. The full extension test suite passes with 68 tests.

File: wntr/extensions/design_verification/constraints.py (hunk 11 above)
Comment (thom1505) [reply wntr/extensions/design_verification/constraints.py]: Addressed. The custom result table validation helper was removed, and the constraint functions now operate directly on the populated WNTR simulation result Data Frames. This simplifies the extension and relies on Simulation Results construction when the simulation completes successfully. The full extension test suite passes with 68 tests.

File: wntr/extensions/design_verification/runner.py (hunk 12 above)
Comment (thom1505) [reply wntr/extensions/design_verification/runner.py]: Addressed. The custom result column selection helper was removed. The runner now relies directly on pandas .loc column selection using WNTR’s junction and pipe name lists, consistent with WNTR’s registry checks for unique node and link names.

File: wntr/extensions/design_verification/models.py (hunk 6 above)
Comment (thom1505) [reply wntr/extensions/design_verification/models.py:58]: Addressed. HydraulicScenario now wraps wntr.network.Options rather than duplicating individual scenario fields. apply_hydraulic_scenario applies the time and hydraulic option groups from the supplied Options object to an independent copy of the water network model. This reuses WNTR’s built-in option validation and substantially simplifies the scenario-handling code, consistent with your suggestion.
